File: src/typeio.c
// clang-format off
#include <postgres.h>
// clang-format on

#include <access/htup_details.h>
#include <catalog/pg_type.h>
#include <fmgr.h>
#include <funcapi.h>
#include <utils/builtins.h>
#include <utils/elog.h>
#include <utils/lsyscache.h>
#include <utils/typcache.h>

#include <caml/alloc.h>
#include <caml/fail.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>

#include "plocaml.h"

/*
 * Value marshalling between PostgreSQL Datums and the OCaml `datum` variant.
 * Input:  make_ocaml_datum / plocaml_build_args  (Datum -> datum)
 * Output: plocaml_convert_datum / *_heap_tuple    (datum -> Datum)
 */

static value plocaml_composite_to_datum(Datum arg);

value make_ocaml_datum(Oid type_oid, Datum val, bool isnull) {
  CAMLparam0();
  CAMLlocal2(v, s);

  if (isnull) {
    CAMLreturn(Val_int(0)); // Null variant
  }

  if (type_oid == INT4OID) {
    v = caml_alloc(1, DATUM_TAG_INT);
    Store_field(v, 0, Val_int(DatumGetInt32(val)));
  } else if (type_oid == INT8OID) {
    v = caml_alloc(1, DATUM_TAG_INT);
    Store_field(v, 0, Val_int(DatumGetInt64(val)));
  } else if (type_oid == INT2OID) {
    v = caml_alloc(1, DATUM_TAG_INT);
    Store_field(v, 0, Val_int(DatumGetInt16(val)));
  } else if (type_oid == FLOAT8OID) {
    v = caml_alloc(1, DATUM_TAG_FLOAT);
    value f = caml_copy_double(DatumGetFloat8(val));
    Store_field(v, 0, f);
  } else if (type_oid == BOOLOID) {
    v = caml_alloc(1, DATUM_TAG_BOOL);
    bool b = DatumGetBool(val);
    Store_field(v, 0, Val_int(b ? 1 : 0));
  } else if (type_is_rowtype(type_oid)) {
    v = plocaml_composite_to_datum(val);
  } else {
    // Default to string for everything else (numeric, enum, uuid, json, ...)
    v = caml_alloc(1, DATUM_TAG_STRING);
    char *str;
    Oid typoutput;
    bool typisvarlena;
    getTypeOutputInfo(type_oid, &typoutput, &typisvarlena);
    str = OidOutputFunctionCall(typoutput, val);
    s = caml_copy_string(str);
    Store_field(v, 0, s);
    pfree(str);
  }
  CAMLreturn(v);
}

/* Convert a composite (row) Datum into a PL.Record: a (column-name, value)
   association list in column order, giving field-by-name access like
   PL/Python's dict (and matching the shape of SPI result rows). */
value plocaml_composite_to_datum(Datum arg) {
  CAMLparam0();
  CAMLlocal5(v, lst, pair, name, fval_v);
  CAMLlocal1(node);

  HeapTupleHeader th = DatumGetHeapTupleHeader(arg);
  Oid tup_type = HeapTupleHeaderGetTypeId(th);
  int32 tup_typmod = HeapTupleHeaderGetTypMod(th);
  TupleDesc tupdesc = lookup_rowtype_tupdesc(tup_type, tup_typmod);

  HeapTupleData tuple;
  tuple.t_len = HeapTupleHeaderGetDatumLength(th);
  ItemPointerSetInvalid(&(tuple.t_self));
  tuple.t_tableOid = InvalidOid;
  tuple.t_data = th;

  /* Build the list backwards so it ends up in column order. */
  lst = Val_int(0); // []
  for (int i = tupdesc->natts - 1; i >= 0; i--) {
    Form_pg_attribute att = TupleDescAttr(tupdesc, i);
    if (att->attisdropped)
      continue;

    bool isnull;
    Datum fval = heap_getattr(&tuple, i + 1, tupdesc, &isnull);
    name = caml_copy_string(NameStr(att->attname));
    fval_v = make_ocaml_datum(att->atttypid, fval, isnull);

    pair = caml_alloc(2, 0);
    Store_field(pair, 0, name);
    Store_field(pair, 1, fval_v);

    node = caml_alloc(2, 0);
    Store_field(node, 0, pair);
    Store_field(node, 1, lst);
    lst = node;
  }
  ReleaseTupleDesc(tupdesc);

  v = caml_alloc(1, DATUM_TAG_RECORD);
  Store_field(v, 0, lst);
  CAMLreturn(v);
}

value plocaml_build_args(FunctionCallInfo fcinfo) {
  CAMLparam0();
  CAMLlocal2(args_arr, v);
  int nargs = PG_NARGS();
  args_arr = caml_alloc(nargs, 0); // 0 tag for Tuple/Array
  for (int i = 0; i < nargs; i++) {
    if (PG_ARGISNULL(i)) {
      v = Val_int(0); // Null variant (integer 0)
    } else {
      Oid type_oid = get_fn_expr_argtype(fcinfo->flinfo, i);
      Datum arg = PG_GETARG_DATUM(i);

      if (type_oid == INT4OID) {
        v = caml_alloc(1, DATUM_TAG_INT);
        Store_field(v, 0, Val_int(DatumGetInt32(arg)));
      } else if (type_oid == FLOAT8OID) {
        v = caml_alloc(1, DATUM_TAG_FLOAT);
        Store_field(v, 0, caml_copy_double(DatumGetFloat8(arg)));
      } else if (type_oid == TEXTOID || type_oid == VARCHAROID) {
        v = caml_alloc(1, DATUM_TAG_STRING);
        char *str = TextDatumGetCString(arg);
        Store_field(v, 0, caml_copy_string(str));
        pfree(str);
      } else if (type_oid == BOOLOID) {
        v = caml_alloc(1, DATUM_TAG_BOOL);
        Store_field(v, 0, Val_int(DatumGetBool(arg) ? 1 : 0));
      } else if (type_is_rowtype(type_oid)) {
        v = plocaml_composite_to_datum(arg);
      } else {
        // Any other type (numeric, enum, uuid, domain, ...) as its text form.
        v = make_ocaml_datum(type_oid, arg, false);
      }
    }
    Store_field(args_arr, i, v);
  }
  CAMLreturn(args_arr);
}

/* Convert one datum value into a column value (Datum + isnull flag). */
static void plocaml_elem_to_datum(value elem, Datum *value_out, bool *isnull) {
  if (Is_long(elem)) {
    *isnull = true;
    *value_out = (Datum)0;
    return;
  }
  *isnull = false;
  switch (Tag_val(elem)) {
  case DATUM_TAG_INT:
    *value_out = Int32GetDatum(Int_val(Field(elem, 0)));
    break;
  case DATUM_TAG_FLOAT:
    *value_out = Float8GetDatum(Double_val(Field(elem, 0)));
    break;
  case DATUM_TAG_STRING:
    *value_out = CStringGetTextDatum(String_val(Field(elem, 0)));
    break;
  case DATUM_TAG_BOOL:
    *value_out = BoolGetDatum(Int_val(Field(elem, 0)) != 0);
    break;
  default:
    elog(ERROR, "PL/OCaml engine error: Unexpected datum variant tag in "
                "composite field.");
  }
}

/* Build a HeapTuple from a PL.Array of field values, positionally. */
static HeapTuple plocaml_build_heap_tuple(value arr, TupleDesc tupdesc) {
  int arr_len = Wosize_val(arr);
  if (arr_len != tupdesc->natts) {
    ereport(ERROR, (errcode(ERRCODE_DATATYPE_MISMATCH),
                    errmsg("length of returned sequence did not match number "
                           "of columns in row")));
  }

  Datum *values = palloc(tupdesc->natts * sizeof(Datum));
  bool *nulls = palloc(tupdesc->natts * sizeof(bool));

  for (int i = 0; i < tupdesc->natts; i++)
    plocaml_elem_to_datum(Field(arr, i), &values[i], &nulls[i]);

  HeapTuple tuple = heap_form_tuple(tupdesc, values, nulls);
  pfree(values);
  pfree(nulls);
  return tuple;
}

/* Build a HeapTuple from a PL.Record ((column-name, value) list), matching each
   result column by name -- the return-side counterpart of a composite argument
   arriving as a Record. */
static HeapTuple plocaml_build_heap_tuple_from_record(value rec_list,
                                                      TupleDesc tupdesc) {
  Datum *values = palloc(tupdesc->natts * sizeof(Datum));
  bool *nulls = palloc(tupdesc->natts * sizeof(bool));

  for (int i = 0; i < tupdesc->natts; i++) {
    Form_pg_attribute att = TupleDescAttr(tupdesc, i);
    if (att->attisdropped) {
      nulls[i] = true;
      values[i] = (Datum)0;
      continue;
    }
    const char *colname = NameStr(att->attname);
    bool found = false;
    for (value node = rec_list; Is_block(node); node = Field(node, 1)) {
      value pair = Field(node, 0);
      if (strcmp(String_val(Field(pair, 0)), colname) == 0) {
        plocaml_elem_to_datum(Field(pair, 1), &values[i], &nulls[i]);
        found = true;
        break;
      }
    }
    if (!found)
      ereport(ERROR,
              (errcode(ERRCODE_DATATYPE_MISMATCH),
               errmsg("record result is missing column \"%s\"", colname)));
  }

  HeapTuple tuple = heap_form_tuple(tupdesc, values, nulls);
  pfree(values);
  pfree(nulls);
  return tuple;
}

/* Build a HeapTuple from a composite datum, which may be a PL.Array
   (positional) or a PL.Record (by column name). */
HeapTuple plocaml_composite_to_heap_tuple(value composite, TupleDesc tupdesc) {
  if (Is_block(composite) && Tag_val(composite) == DATUM_TAG_ARRAY)
    return plocaml_build_heap_tuple(Field(composite, 0), tupdesc);
  if (Is_block(composite) && Tag_val(composite) == DATUM_TAG_RECORD)
    return plocaml_build_heap_tuple_from_record(Field(composite, 0), tupdesc);
  ereport(ERROR, (errcode(ERRCODE_DATATYPE_MISMATCH),
                  errmsg("composite result must be a PL.Array or PL.Record")));
  return NULL; /* unreachable */
}

Datum plocaml_convert_datum(FunctionCallInfo fcinfo, value datum_val,
                            bool *isnull) {
  if (Is_long(datum_val)) {
    *isnull = true;
    return (Datum)0;
  }

  *isnull = false;
  int d_tag = Tag_val(datum_val);

  if (d_tag == DATUM_TAG_INT) {
    return Int32GetDatum(Int_val(Field(datum_val, 0)));
  } else if (d_tag == DATUM_TAG_FLOAT) {
    return Float8GetDatum(Double_val(Field(datum_val, 0)));
  } else if (d_tag == DATUM_TAG_STRING) {
    char *str = (char *)String_val(Field(datum_val, 0));
    /* A string returned for a composite result is parsed via the row type's
       input function, mirroring PL/Python's PLyUnicode_ToComposite (e.g.
       returning "(1,foo)" for a composite type). */
    Oid rettype = get_fn_expr_rettype(fcinfo->flinfo);
    if (OidIsValid(rettype) && type_is_rowtype(rettype)) {
      Oid in_fn, ioparam;
      getTypeInputInfo(rettype, &in_fn, &ioparam);
      return OidInputFunctionCall(in_fn, str, ioparam, -1);
    }
    return CStringGetTextDatum(str);
  } else if (d_tag == DATUM_TAG_BOOL) {
    return BoolGetDatum(Int_val(Field(datum_val, 0)) != 0);
  } else if (d_tag == DATUM_TAG_ARRAY || d_tag == DATUM_TAG_RECORD) {
    TupleDesc tupdesc;
    if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE) {
      ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                      errmsg("function returning record called in context that "
                             "cannot accept type record")));
    }
    BlessTupleDesc(tupdesc);
    HeapTuple tuple = plocaml_composite_to_heap_tuple(datum_val, tupdesc);
    Datum return_datum = heap_copy_tuple_as_datum(tuple, tupdesc);
    heap_freetuple(tuple);
    return return_datum;
  } else {
    elog(ERROR, "PL/OCaml engine error: Unexpected datum variant tag.");
  }
  return (Datum)0;
}

value plocaml_heap_tuple_to_record(HeapTuple tuple, TupleDesc tupdesc) {
  CAMLparam0();
  CAMLlocal4(row_list, pair, col_name, col_val);

  row_list = Val_int(0); // []

  // Build the list backwards so it ends up in the correct order
  for (int j = tupdesc->natts; j > 0; j--) {
    Form_pg_attribute att = TupleDescAttr(tupdesc, j - 1);
    if (att->attisdropped)
      continue;

    bool isnull;
    Datum val = heap_getattr(tuple, j, tupdesc, &isnull);
    Oid type_oid = att->atttypid;
    char *fname = NameStr(att->attname);

    col_name = caml_copy_string(fname);
    col_val = make_ocaml_datum(type_oid, val, isnull);

    pair = caml_alloc(2, 0);
    Store_field(pair, 0, col_name);
    Store_field(pair, 1, col_val);

    value new_node = caml_alloc(2, 0);
    Store_field(new_node, 0, pair);
    Store_field(new_node, 1, row_list);
    row_list = new_node;
  }

  value record_val = caml_alloc(1, DATUM_TAG_RECORD);
  Store_field(record_val, 0, row_list);
  CAMLreturn(record_val);
}

File: src/typeio.rs
use pgrx::pg_sys;
use pgrx::PgTupleDesc;
use std::ffi::CStr;

const DATUM_TAG_INT: u8 = 0;
const DATUM_TAG_FLOAT: u8 = 1;
const DATUM_TAG_STRING: u8 = 2;
const DATUM_TAG_BOOL: u8 = 3;
#[allow(dead_code)]
const DATUM_TAG_ARRAY: u8 = 4;
const DATUM_TAG_RECORD: u8 = 5;

unsafe fn make_ocaml_null() -> ocaml::Value {
    ocaml::Value::new(ocaml::sys::val_int(0))
}

unsafe fn make_ocaml_int(i: isize) -> ocaml::Value {
    let block = ocaml::sys::caml_alloc(1, DATUM_TAG_INT);
    *ocaml::sys::field(block, 0) = ocaml::sys::val_int(i);
    ocaml::Value::new(block)
}

unsafe fn make_ocaml_float(f: f64) -> ocaml::Value {
    let d = ocaml::sys::caml_copy_double(f);
    let block = ocaml::sys::caml_alloc(1, DATUM_TAG_FLOAT);
    *ocaml::sys::field(block, 0) = d;
    ocaml::Value::new(block)
}

unsafe fn make_ocaml_string(s: &str) -> ocaml::Value {
    let str_val = ocaml::Value::string(s);
    let block = ocaml::sys::caml_alloc(1, DATUM_TAG_STRING);
    *ocaml::sys::field(block, 0) = str_val.raw().0;
    ocaml::Value::new(block)
}

unsafe fn make_ocaml_bool(b: bool) -> ocaml::Value {
    let block = ocaml::sys::caml_alloc(1, DATUM_TAG_BOOL);
    *ocaml::sys::field(block, 0) = ocaml::sys::val_int(if b { 1 } else { 0 });
    ocaml::Value::new(block)
}

unsafe fn composite_to_ocaml_datum(arg: pg_sys::Datum) -> ocaml::Value {
    let th = pg_sys::pg_detoast_datum(arg.cast_mut_ptr()) as pg_sys::HeapTupleHeader;
    let tup_type = pgrx::htup::heap_tuple_header_get_type_id(th);
    let tup_typmod = pgrx::htup::heap_tuple_header_get_typmod(th);
    let tupdesc = pg_sys::lookup_rowtype_tupdesc(tup_type, tup_typmod);

    let mut tuple = pg_sys::HeapTupleData {
        t_len: pgrx::htup::heap_tuple_header_get_datum_length(th) as u32,
        t_self: pg_sys::ItemPointerData::default(),
        t_tableOid: pg_sys::InvalidOid,
        t_data: th,
    };

    let list = heap_tuple_to_row_list(&mut tuple, tupdesc);
    pgrx::tupdesc::release_tupdesc(tupdesc);

    let record_block = ocaml::sys::caml_alloc(1, DATUM_TAG_RECORD);
    *ocaml::sys::field(record_block, 0) = list.raw().0;
    ocaml::Value::new(record_block)
}

pub(crate) unsafe fn make_ocaml_datum(
    type_oid: pg_sys::Oid,
    val: pg_sys::Datum,
    isnull: bool,
) -> ocaml::Value {
    if isnull {
        return make_ocaml_null();
    }

    if type_oid == pg_sys::INT2OID {
        let i = pg_sys::DatumGetInt16(val) as isize;
        make_ocaml_int(i)
    } else if type_oid == pg_sys::INT4OID {
        let i = pg_sys::DatumGetInt32(val) as isize;
        make_ocaml_int(i)
    } else if type_oid == pg_sys::INT8OID {
        let i = pg_sys::DatumGetInt64(val) as isize;
        make_ocaml_int(i)
    } else if type_oid == pg_sys::FLOAT4OID {
        let f = pg_sys::DatumGetFloat4(val) as f64;
        make_ocaml_float(f)
    } else if type_oid == pg_sys::FLOAT8OID {
        let f = pg_sys::DatumGetFloat8(val);
        make_ocaml_float(f)
    } else if type_oid == pg_sys::BOOLOID {
        let b = pg_sys::DatumGetBool(val);
        make_ocaml_bool(b)
    } else if pg_sys::type_is_rowtype(type_oid) {
        composite_to_ocaml_datum(val)
    } else {
        let mut typoutput = pg_sys::InvalidOid;
        let mut typisvarlena = false;
        pg_sys::getTypeOutputInfo(type_oid, &mut typoutput, &mut typisvarlena);
        let cstr_ptr = pg_sys::OidOutputFunctionCall(typoutput, val);
        let s = CStr::from_ptr(cstr_ptr).to_str().unwrap_or("");
        let ocaml_datum = make_ocaml_string(s);
        pg_sys::pfree(cstr_ptr.cast());
        ocaml_datum
    }
}

pub(crate) unsafe fn heap_tuple_to_row_list(
    tuple: pg_sys::HeapTuple,
    tupdesc_ptr: pg_sys::TupleDesc,
) -> ocaml::Value {
    let pg_tupdesc = PgTupleDesc::from_pg_unchecked(tupdesc_ptr);
    let natts = pg_tupdesc.len();
    let mut list = ocaml::Value::new(ocaml::sys::val_int(0)); // []

    for j in (0..natts).rev() {
        let att = match pg_tupdesc.get(j) {
            Some(a) => a,
            None => continue,
        };

        if att.attisdropped {
            continue;
        }

        let mut isnull = false;
        let val = pg_sys::heap_getattr(tuple, (j + 1) as i32, tupdesc_ptr, &mut isnull);
        let type_oid = att.atttypid;
        let col_name_cstr = CStr::from_ptr(att.attname.data.as_ptr());
        let col_name = col_name_cstr.to_str().unwrap_or("");

        let name_val = ocaml::Value::string(col_name);
        let datum_val = make_ocaml_datum(type_oid, val, isnull);

        let pair = ocaml::sys::caml_alloc(2, 0);
        *ocaml::sys::field(pair, 0) = name_val.raw().0;
        *ocaml::sys::field(pair, 1) = datum_val.raw().0;
        let pair_val = ocaml::Value::new(pair);

        let node = ocaml::sys::caml_alloc(2, 0);
        *ocaml::sys::field(node, 0) = pair_val.raw().0;
        *ocaml::sys::field(node, 1) = list.raw().0;
        list = ocaml::Value::new(node);
    }

    list
}

/// Converts a string into PostgreSQL's internal binary Datum representation
/// for `type_oid` by invoking the type's registered input function (`typinput`).
///
/// This serves as the counterpart to `OidOutputFunctionCall` used in `make_ocaml_datum`,
/// allowing arbitrary PostgreSQL types (such as `numeric`, `timestamp`, `uuid`, `jsonb`, etc.)
/// to be parsed from string representations when passing parameters to prepared SPI plans.
pub(crate) unsafe fn string_to_pg_datum(
    s: &str,
    type_oid: pg_sys::Oid,
) -> Result<pg_sys::Datum, String> {
    let mut typinput = pg_sys::InvalidOid;
    let mut typioparam = pg_sys::InvalidOid;
    pg_sys::getTypeInputInfo(type_oid, &mut typinput, &mut typioparam);
    let cstr =
        std::ffi::CString::new(s).map_err(|_| "String argument contains null byte".to_string())?;
    let datum = pg_sys::OidInputFunctionCall(typinput, cstr.as_ptr().cast_mut(), typioparam, -1);
    Ok(datum)
}

pub(crate) unsafe fn ocaml_datum_to_pg_datum(
    val: ocaml::sys::Value,
    type_oid: pg_sys::Oid,
) -> Result<(pg_sys::Datum, bool), String> {
    if ocaml::sys::is_long(val) {
        // Tag Null (int 0)
        return Ok((pg_sys::Datum::from(0), true));
    }

    let tag = ocaml::sys::tag_val(val);
    let datum = match tag {
        DATUM_TAG_INT => {
            let n = ocaml::sys::int_val(*ocaml::sys::field(val, 0));
            if type_oid == pg_sys::INT2OID {
                pg_sys::Int16GetDatum(n as i16)
            } else if type_oid == pg_sys::INT4OID {
                pg_sys::Int32GetDatum(n as i32)
            } else if type_oid == pg_sys::INT8OID {
                pg_sys::Int64GetDatum(n as i64)
            } else if type_oid == pg_sys::FLOAT4OID {
                pg_sys::Float4GetDatum(n as f32)
            } else if type_oid == pg_sys::FLOAT8OID {
                pg_sys::Float8GetDatum(n as f64)
            } else if type_oid == pg_sys::BOOLOID {
                pg_sys::BoolGetDatum(n != 0)
            } else {
                string_to_pg_datum(&n.to_string(), type_oid)?
            }
        }
        DATUM_TAG_FLOAT => {
            let f = ocaml::Value::new(*ocaml::sys::field(val, 0)).double_val();
            if type_oid == pg_sys::FLOAT4OID {
                pg_sys::Float4GetDatum(f as f32)
            } else if type_oid == pg_sys::FLOAT8OID {
                pg_sys::Float8GetDatum(f)
            } else if type_oid == pg_sys::INT2OID {
                pg_sys::Int16GetDatum(f as i16)
            } else if type_oid == pg_sys::INT4OID {
                pg_sys::Int32GetDatum(f as i32)
            } else if type_oid == pg_sys::INT8OID {
                pg_sys::Int64GetDatum(f as i64)
            } else {
                string_to_pg_datum(&f.to_string(), type_oid)?
            }
        }
        DATUM_TAG_STRING => {
            let s_val = ocaml::Value::new(*ocaml::sys::field(val, 0));
            let s: &str = ocaml::FromValue::from_value(s_val);
            string_to_pg_datum(s, type_oid)?
        }
        DATUM_TAG_BOOL => {
            let b = ocaml::sys::int_val(*ocaml::sys::field(val, 0)) != 0;
            if type_oid == pg_sys::BOOLOID {
                pg_sys::BoolGetDatum(b)
            } else if type_oid == pg_sys::INT4OID {
                pg_sys::Int32GetDatum(if b { 1 } else { 0 })
            } else {
                string_to_pg_datum(if b { "true" } else { "false" }, type_oid)?
            }
        }
        _ => {
            return Err(format!("Unsupported datum tag {tag} for SPI argument"));
        }
    };

    Ok((datum, false))
}

pub(crate) unsafe fn ocaml_value_to_pg_datum(
    val: ocaml::sys::Value,
    type_oid: pg_sys::Oid,
) -> Result<(pg_sys::Datum, bool), String> {
    if type_oid == pg_sys::VOIDOID {
        return Ok((pg_sys::Datum::from(0), false));
    }

    if ocaml::sys::is_long(val) {
        let n = ocaml::sys::int_val(val);
        if type_oid == pg_sys::BOOLOID {
            return Ok((pg_sys::BoolGetDatum(n != 0), false));
        } else if type_oid == pg_sys::INT2OID {
            return Ok((pg_sys::Int16GetDatum(n as i16), false));
        } else if type_oid == pg_sys::INT4OID {
            return Ok((pg_sys::Int32GetDatum(n as i32), false));
        } else if type_oid == pg_sys::INT8OID {
            return Ok((pg_sys::Int64GetDatum(n as i64), false));
        } else if type_oid == pg_sys::FLOAT4OID {
            return Ok((pg_sys::Float4GetDatum(n as f32), false));
        } else if type_oid == pg_sys::FLOAT8OID {
            return Ok((pg_sys::Float8GetDatum(n as f64), false));
        } else if n == 0 {
            // Null or None
            return Ok((pg_sys::Datum::from(0), true));
        } else {
            let datum = string_to_pg_datum(&n.to_string(), type_oid)?;
            return Ok((datum, false));
        }
    }

    let tag = ocaml::sys::tag_val(val);
    let size = ocaml::sys::wosize_val(val);

    if tag == ocaml::sys::STRING {
        let s_val = ocaml::Value::new(val);
        let s: &str = ocaml::FromValue::from_value(s_val);
        let datum = string_to_pg_datum(s, type_oid)?;
        return Ok((datum, false));
    }

    if tag == ocaml::sys::DOUBLE {
        let f = ocaml::Value::new(val).double_val();
        if type_oid == pg_sys::FLOAT4OID {
            return Ok((pg_sys::Float4GetDatum(f as f32), false));
        } else if type_oid == pg_sys::FLOAT8OID {
            return Ok((pg_sys::Float8GetDatum(f), false));
        } else if type_oid == pg_sys::INT2OID {
            return Ok((pg_sys::Int16GetDatum(f as i16), false));
        } else if type_oid == pg_sys::INT4OID {
            return Ok((pg_sys::Int32GetDatum(f as i32), false));
        } else if type_oid == pg_sys::INT8OID {
            return Ok((pg_sys::Int64GetDatum(f as i64), false));
        } else {
            let datum = string_to_pg_datum(&f.to_string(), type_oid)?;
            return Ok((datum, false));
        }
    }

    match tag {
        DATUM_TAG_INT if size == 1 => {
            let field0 = *ocaml::sys::field(val, 0);
            if ocaml::sys::is_long(field0) {
                let n = ocaml::sys::int_val(field0);
                if type_oid == pg_sys::INT2OID {
                    return Ok((pg_sys::Int16GetDatum(n as i16), false));
                } else if type_oid == pg_sys::INT4OID {
                    return Ok((pg_sys::Int32GetDatum(n as i32), false));
                } else if type_oid == pg_sys::INT8OID {
                    return Ok((pg_sys::Int64GetDatum(n as i64), false));
                } else if type_oid == pg_sys::FLOAT4OID {
                    return Ok((pg_sys::Float4GetDatum(n as f32), false));
                } else if type_oid == pg_sys::FLOAT8OID {
                    return Ok((pg_sys::Float8GetDatum(n as f64), false));
                } else if type_oid == pg_sys::BOOLOID {
                    return Ok((pg_sys::BoolGetDatum(n != 0), false));
                } else {
                    let datum = string_to_pg_datum(&n.to_string(), type_oid)?;
                    return Ok((datum, false));
                }
            } else {
                return ocaml_value_to_pg_datum(field0, type_oid);
            }
        }
        DATUM_TAG_FLOAT if size == 1 => {
            let f = ocaml::Value::new(*ocaml::sys::field(val, 0)).double_val();
            if type_oid == pg_sys::FLOAT4OID {
                return Ok((pg_sys::Float4GetDatum(f as f32), false));
            } else if type_oid == pg_sys::FLOAT8OID {
                return Ok((pg_sys::Float8GetDatum(f), false));
            } else if type_oid == pg_sys::INT2OID {
                return Ok((pg_sys::Int16GetDatum(f as i16), false));
            } else if type_oid == pg_sys::INT4OID {
                return Ok((pg_sys::Int32GetDatum(f as i32), false));
            } else if type_oid == pg_sys::INT8OID {
                return Ok((pg_sys::Int64GetDatum(f as i64), false));
            } else {
                let datum = string_to_pg_datum(&f.to_string(), type_oid)?;
                return Ok((datum, false));
            }
        }
        DATUM_TAG_STRING if size == 1 => {
            let s_val = ocaml::Value::new(*ocaml::sys::field(val, 0));
            let s: &str = ocaml::FromValue::from_value(s_val);
            let datum = string_to_pg_datum(s, type_oid)?;
            return Ok((datum, false));
        }
        DATUM_TAG_BOOL if size == 1 => {
            let b = ocaml::sys::int_val(*ocaml::sys::field(val, 0)) != 0;
            if type_oid == pg_sys::BOOLOID {
                return Ok((pg_sys::BoolGetDatum(b), false));
            } else if type_oid == pg_sys::INT4OID {
                return Ok((pg_sys::Int32GetDatum(if b { 1 } else { 0 }), false));
            } else {
                let datum = string_to_pg_datum(if b { "true" } else { "false" }, type_oid)?;
                return Ok((datum, false));
            }
        }
        _ => {
            return Err(format!(
                "Unsupported return value tag {tag} for PostgreSQL type {type_oid:?}"
            ));
        }
    }
}
